proposals/interface-types/working-notes/scenarios/chooser.md

```diff
@@ -0,0 +1,680 @@
+  (param $l i32)
+  (param $r i32)
+  (result i32)
+  (local $o1 i32) ;; left string shared-ptr memory
+  (local $o2 i32) ;; right string shared-ptr memory
+  (local $o3 i32)
+
+  local.get $l
+  call-export "access-utf8-i" ;; return base & len
+  
+  let (local $lbase i32)(local $lsize i32)(result i32)
+    local.get $lbase
+    local.get $lsize
+    call-export "malloc-x"
+    let (local $ptr i32)(result i32)
+	  local.get $lbase
+	  local.get $ptr
+	  local.get $lsize
+      memory.copy "mem-i" "mem-x" ;; copy string across
+      local.get $ptr
+      i32.const 1                ;; initial reference count of 1
+      call-export "shared-builder-x" ;; make it a shared ptr pair
+      local.tee $o1
+    end
+    let (local $lp i32) (result i32)
+      local.get $r
+      call-export "access-utf8-i"
+      let (local $rbase i32)(local $rsize i32)(result i32)
+        local.get $rbase
+        local.get $rsize
+        call-export "malloc-x"
+	let (local $ptr i32)(result i32)
+          local.get $rbase
+	  local.get $rsize
+          local.get $ptr
+	  local.get $rsize
+          memory.copy "mem-i" "mem-x" ;; copy string across
+          local.get $ptr
+          i32.const 1                ;; initial reference count of 1
+          call-export "shared-builder-x" ;; make it a shared ptr pair
+	  local.tee $o2
+        end
+        let (local $rp i32)(result i32)
+          local.get $lp
+          local.get $rp
+          call-export "nondeterministic_choice"  ;; call the chooser itself
+	  local.tee $o3
+          call-export "shared-get-x"
+          call-export "access-utf8-x"
+          let (local $xbase i32) (local $xsize i32) (result i32)
+            local.get $xsize
+            call-export "malloc-i"     ;; create string entity of right size
+            let (local $ptr i32)
+              local.get $xbase
+              local.get $ptr
+              local.get $xsize
+              memory.copy "mem-x" "mem-i"
+              local.get $ptr           ;; our final return value
+            end
+          end
+        end
+      end
+    end
+  end
+  local.get $o1
+  call-export "shared-release-x"
+  local.get $o2
+  call-export "shared-release-x"
+  local.get $o3
+  call-export "shared-release-x" 
+)
+```
+
+For simplicity, we migrated all the deallocations to the end of the fused
+adapter. In some situations, for example when processing arrays, we may wish to
+be more aggressive in invoking the memory release code.
+
+Although fairly long, this fused adapter has a striaghtfoward structure: the
+input strings are copied from the import memory to the export memory -- and also
+wrapped as shared pointer structures as required by the signature of the
+`nondeterministic_chooser` function. The resulting string is copied from the
+export localtion to the import memory. And, finally, any temporary structures
+allocated are released.
+
+Note that the string returned by `nondeterministic_chooser` is not directly
```

Comment:
I think I understand what this is getting at, but the sentence doesn't parse for me

Reply:
reworded some. Missing text or over-enthusiastic use of delete ...